lang/driver/src/database.rs

```diff
@@ -41,6 +44,204 @@ pub struct Database {
+        }
+    }
+
+    fn recompute_source(&mut self, uri: &Url) -> Result<String, Error> {
+        let source = self.source.read_to_string(uri)?;
+        let file = codespan::File::new(uri.as_str().into(), source.clone());
+        self.files.insert(uri.clone(), file);
+        Ok(source)
+    }
+
+    // Core API: CST (Concrete Syntax Tree)
+    //
+    //
+
+    pub fn cst(&mut self, uri: &Url) -> Result<Arc<cst::decls::Module>, Error> {
+        match self.cst.get_unless_stale(uri) {
+            Some(cst) => cst.clone(),
+            None => self.recompute_cst(uri),
+        }
+    }
+
+    fn recompute_cst(&mut self, uri: &Url) -> Result<Arc<cst::decls::Module>, Error> {
+        let source = self.source(uri)?;
+        log::debug!("Parsing module: {}", uri);
+        let module =
+            parser::parse_module(uri.clone(), &source).map_err(Error::Parser).map(Arc::new);
+        self.cst.insert(uri.clone(), module.clone());
+        module
+    }
+
+    // Core API: SymbolTable
+    //
+    //
+
+    pub fn symbol_table(&mut self, uri: &Url) -> Result<ModuleSymbolTable, Error> {
+        match self.symbol_table.get_unless_stale(uri) {
+            Some(symbol_table) => Ok(symbol_table.clone()),
+            None => self.recompute_symbol_table(uri),
+        }
+    }
+
+    fn recompute_symbol_table(&mut self, uri: &Url) -> Result<ModuleSymbolTable, Error> {
+        let cst = self.cst(uri)?;
+        let module_symbol_table = lowering::build_symbol_table(&cst)?;
+        self.symbol_table.insert(uri.clone(), module_symbol_table.clone());
+        Ok(module_symbol_table)
+    }
+
+    // Core API: UST
+    //
+    //
+
+    pub fn ust(&mut self, uri: &Url) -> Result<Arc<ast::Module>, Error> {
+        match self.ust.get_unless_stale(uri) {
+            Some(ust) => ust.clone(),
+            None => self.recompute_ust(uri),
+        }
+    }
+
+    pub fn recompute_ust(&mut self, uri: &Url) -> Result<Arc<ast::Module>, Error> {
+        let cst = self.cst(uri)?;
+        let deps = self
+            .deps
+            .get(uri)
+            .ok_or(Error::Driver(DriverError::Impossible(format!("Did not find deps for {}", uri))))
+            .cloned()?;
+
+        // Compute the SymbolTable consisting of all the
+        // ModuleSymbolTables of all direct dependencies
+        // and the SymbolTable from the module itself.
+        let mut symbol_table = SymbolTable::default();
+        let module_symbol_table = self.symbol_table(&uri)?;
+        symbol_table.insert(uri.clone(), module_symbol_table);
+        for dep in deps {
+            let module_symbol_table = self.symbol_table(&dep)?;
+            symbol_table.insert(dep.clone(), module_symbol_table);
+        }
+
+        let ust = lowering::lower_module_with_symbol_table(&cst, &symbol_table)
+            .map_err(Error::Lowering)
+            .map(Arc::new);
+
+        self.ust.insert(uri.clone(), ust.clone());
+        ust
+    }
+
+    // Core API: AST
+    //
+    // TODO
```

Comment:
TODO

Reply:
Ah yes, that is my marker that the three functions below haven't been ported to the new style with `ast()` and `recompute_ast()` yet.